libraries/AP_OpenDroneID/AP_OpenDroneID.cpp

```diff
@@ -0,0 +1,607 @@
+}
+
+void AP_OpenDroneID::send_static_out()
+{
+    const uint32_t now_ms = AP_HAL::millis();
+
+    // we need to notify user if we lost the transmitter
+    if (now_ms - last_arm_status_ms > 5000) {
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "ODID: lost transmitter");
+    }
+
+    const uint32_t msg_spacing_ms = 5000 / 4;
+    if (now_ms - last_msg_send_ms >= msg_spacing_ms) {
+        // allow update of channel during setup, this makes it easy to debug with a GCS
+        _chan = mavlink_channel_t(gcs().get_channel_from_port_number(_mav_port));
+
+        last_msg_send_ms = now_ms;
+        switch (next_msg_to_send) {
+        case NEXT_MSG_BASIC_ID:
+            send_basic_id_message();
+            break;
+        case NEXT_MSG_SYSTEM:
+            send_system_message();
+            break;
+        case NEXT_MSG_SELF_ID:
+            send_self_id_message();
+            break;
+        case NEXT_MSG_OPERATOR_ID:
+            send_operator_id_message();
+            break;
+        case NEXT_MSG_ENUM_END:
+            break;
+        }
+        next_msg_to_send = next_msg((uint8_t(next_msg_to_send) + 1) % uint8_t(NEXT_MSG_ENUM_END));
+    }
+}
+
+// The send_location_message
+// all open_drone_id send functions use data stored in the open drone id class.
+//This location send function is an exception. It uses live location data from the ArduPilot system.
+void AP_OpenDroneID::send_location_message()
+{
+    AP_AHRS &ahrs = AP::ahrs();
+    AP_Baro &barometer = AP::baro();
+    const AP_GPS &gps = AP::gps();
+
+    AP_GPS::GPS_Status gps_status = gps.status();
+    const bool got_bad_gps_fix = (gps_status < AP_GPS::GPS_Status::GPS_OK_FIX_3D);
+
+    Location current_location;
+    if (!ahrs.get_location(current_location)) {
+        return;
+    }
+    const uint8_t uav_status = hal.util->get_soft_armed()? MAV_ODID_STATUS_AIRBORNE : MAV_ODID_STATUS_GROUND;
+
+    // What are the conditions where we have invalid groundspeed vector?
+    // For GPS it isn't very accurate if we aren't moving...?
+
+    float direction = ODID_INV_DIR;
+    if (!got_bad_gps_fix) {
+        direction = wrap_360(degrees(ahrs.groundspeed_vector().angle())); // heading (degrees)
+    }
+
+    const float speed_horizontal = create_speed_horizontal(ahrs.groundspeed());
+
+    Vector3f velNED;
+    UNUSED_RESULT(ahrs.get_velocity_NED(velNED));
+    const float climb_rate = create_speed_vertical(-velNED.z); //make sure climb_rate is within Remote ID limit
+
+    int32_t latitude = 0;
+    int32_t longitude = 0;
+    if (current_location.check_latlng()) { //set location if they are valid
+        latitude = current_location.lat;
+        longitude = current_location.lng;
+    }
+
+    // FIXMME
+    // Q? The mavlink docs reference being calculated against  29.92 in HG. Presumably it is actually against the US Standard Atmosphere Standard Day
+    // need to be sure AP is outputting the correct barometric altitude frame here pressure and temperature assumptions.
```

Comment:
The ASTM spec define it as "Pressure Altitude (Ref 29.92 inHg, 1013.24 mb)" and the ASD-STAN spec as "Pressure Altitude (101,325 kPa)"

Reply:
fixed